docs/dev/topology-over-raft.md

```diff
@@ -362,3 +368,75 @@ In case of failures like timeouts, connection issues, etc., the topology
 coordinator keeps retrying. Eventually, it can give up and just move the new
 node to the `left` state, either due to a timeout or an operator intervention,
 but this is not implemented yet.
+
+# Choosing how to join a cluster
+
+A bootstrapping node needs to determine whether to use raft topology operations
+or not. In order to do so, group 0 discovery is performed which establishes
+whether the node is supposed to create a new cluster or join an existing one:
+
+- If the node creates a new cluster, it bootstraps in raft topology mode.
+- If the node joins an existing cluster, it learns about a node which is already
+  a part of the group 0. Such a node will already know whether the cluster
+  operates in raft topology mode or not. It then issues `JOIN_NODE_QUERY`
+  RPC which tells the new node how it should join the cluster.
+
+# Upgrade from legacy topology to raft-based topology
+
+Upgrading existing clusters to use raft-based topology operations is supported.
+The process is not fully automatic because it needs to be triggered by the administrator
+after making sure that no topology operations are running at the moment. This
+limitation might be lifted in the future, and the procedure will trigger
+automatically after making sure that it's safe to do so.
+
+From the admin's point of view, the steps are as follows:
+
+- Upgrade all nodes in the cluster to the newest version
+- Wait until the `SUPPORTS_CONSISTENT_TOPOLOGY_CHANGES` feature becomes enabled
+  and an appropriate message about the cluster being ready for upgrade is printed
+- Trigger the upgrade via `POST /storage_service/raft_topology/upgrade` HTTP route
+- Monitor progress of the upgrade via `GET /storage_service/raft_topology/upgrade`
+  or via observing the logs
```

Comment:
> Maybe the POST should block until upgrade is finished?

Perhaps. Should it wait until the current node reaches `done`, or all nodes reach `done`? As currently described in `topology-on-raft.md`, it is necessary to check that all nodes reached `done` state, otherwise if you only check one then some nodes might still be in `final_global_barrier`.

> Although I'm guessing this also waits for the group0/schema part of the upgrade procedure (if applicable) (I recall we briefly discussed this at some point)

Yes, it does. This is the first thing that `build_coordinator_state` does.

Reply:
Gleb's arguments in the other thread convinced me that it's not worth implementing the blocking semantics for `upgrade` (needs an additional step), so I'm resolving.